core/arch/aarch64/clean_call_opt.c

```diff
@@ -38,6 +38,117 @@
+ * Instead we test the encoding directly.
+ */
+
+/* remove variable bits in the encoding */
+#define STP_LDP_ENC_MASK 0x7fc07fff
+#define STR_LDR_ENC_MASK 0xbfc003ff
+#define MOV_STK_ENC_MASK 0x7f0003ff
+#define STP_LDP_REG_MASK 0xffff83e0
+#define STR_LDR_REG_MASK 0xffffffe0
+
+/* stp x29, x30, [sp, #frame_size]! */
+#define PUSH_FP_LR_ENC 0x29807bfd
+/* ldp x29, x30, [sp], #frame_size */
+#define POP_FP_LR_ENC 0x28c07bfd
+/* add sp, sp, #frame_size */
+#define ADD_SP_ENC 0x110003ff
+/* sub sp, sp, #frame_size */
+#define SUB_SP_ENC 0x510003ff
+/* mov x29, sp */
+#define MOV_X29_SP_ENC 0x910003fd
+/* stp xx, xx, [sp, #offset] */
+#define STP_SP_ENC 0x290003e0
+/* ldp xx, xx, [sp, #offset] */
+#define LDP_SP_ENC 0x294003e0
+/* str xx, [sp, #offset] */
+#define STR_SP_ENC 0xb90003e0
+/* ldr xx, [sp, #offset] */
+#define LDR_SP_ENC 0xb94003e0
+
+static inline bool
+instr_is_push_fp_and_lr(instr_t *instr)
+{
+    uint enc = *(uint *)instr->bytes;
+    return (enc & STP_LDP_ENC_MASK) == PUSH_FP_LR_ENC;
+}
+
+static inline bool
+instr_is_pop_fp_and_lr(instr_t *instr)
+{
+    uint enc = *(uint *)instr->bytes;
+    return (enc & STP_LDP_ENC_MASK) == POP_FP_LR_ENC;
+}
+
+static inline bool
+instr_is_move_frame_ptr(instr_t *instr)
+{
+    uint enc = *(uint *)instr->bytes;
+    return enc == MOV_X29_SP_ENC;
+}
+
+static inline bool
+instr_is_add_stk_ptr(instr_t *instr)
+{
+    uint enc = *(uint *)instr->bytes;
+    return (enc & MOV_STK_ENC_MASK) == ADD_SP_ENC;
+}
+
+static inline bool
+instr_is_sub_stk_ptr(instr_t *instr)
+{
+    uint enc = *(uint *)instr->bytes;
+    return (enc & MOV_STK_ENC_MASK) == SUB_SP_ENC;
+}
+
+static inline bool
+instr_is_push_reg_pair(instr_t *instr, reg_id_t *reg1, reg_id_t *reg2)
+{
+    uint enc = *(uint *)instr->bytes;
+    enc = enc & STP_LDP_ENC_MASK;
+    if ((enc & STP_LDP_REG_MASK) != STP_SP_ENC)
+        return false;
+    *reg1 = (reg_id_t)(enc & 31) + DR_REG_START_GPR;
+    *reg2 = (reg_id_t)(enc >> 10 & 31) + DR_REG_START_GPR;
+    return true;
+}
+
+static inline bool
+instr_is_pop_reg_pair(instr_t *instr, reg_id_t *reg1, reg_id_t *reg2)
+{
+    uint enc = *(uint *)instr->bytes;
+    enc = enc & STP_LDP_ENC_MASK;
+    if ((enc & STP_LDP_REG_MASK) != LDP_SP_ENC)
+        return false;
+    *reg1 = (reg_id_t)(enc & 31) + DR_REG_START_GPR;
+    *reg2 = (reg_id_t)(enc >> 10 & 31) + DR_REG_START_GPR;
+    return true;
+}
+
+static inline bool
+instr_is_push_reg(instr_t *instr, reg_id_t *reg)
+{
+    uint enc = *(uint *)instr->bytes;
+    enc = enc & STR_LDR_ENC_MASK;
+    if ((enc & STR_LDR_REG_MASK) != STR_SP_ENC)
+        return false;
+    *reg = (reg_id_t)(enc & 31) + DR_REG_START_GPR;
+    return true;
+}
+
+static inline bool
+instr_is_pop_reg(instr_t *instr, reg_id_t *reg)
+{
+    uint enc = *(uint *)instr->bytes;
+    enc = enc & STR_LDR_ENC_MASK;
+    if ((enc & STR_LDR_REG_MASK) != LDR_SP_ENC)
+        return false;
+    *reg = (reg_id_t)(enc & 31) + DR_REG_START_GPR;
+    return true;
+}
+
```

Comment:
The client.inline test in inline.dll.c tests a bunch of specific instruction sequences in clean callees for x86.  It would be good to add something similar for aarch.